sleap/gui/commands.py

```diff
@@ -1488,30 +1489,106 @@ def do_action(context: CommandContext, params: dict):
 
     @staticmethod
-    def ask(context: CommandContext, params: dict) -> bool:
+    def ask_and_do(context: CommandContext, params: dict) -> bool:
         """Shows gui for replacing videos in project."""
-        paths = [video.backend.filename for video in context.labels.videos]
 
-        okay = MissingFilesDialog(filenames=paths, replace=True).exec_()
+        # Warn user: newly added labels will be discarded if project is not saved
+        if not context.state["filename"] or context.state["has_changes"]:
+            QMessageBox(
+                text=("You have unsaved changes. Please save before replacing videos.")
+            ).exec_()
+            return False
 
+        # Select the videos we want to swap
+        old_paths = [video.backend.filename for video in context.labels.videos]
+        paths = list(old_paths)
+        okay = MissingFilesDialog(filenames=paths, replace=True).exec_()
         if not okay:
             return False
 
-        params["new_video_paths"] = paths
+        # Only return an import list for videos we swap
+        new_paths = [
+            (path, video_idx)
+            for video_idx, (path, old_path) in enumerate(zip(paths, old_paths))
+            if path != old_path
+        ]
+
+        new_paths = []
+        old_videos = []
+        all_videos = context.labels.videos
+        for video_idx, (path, old_path) in enumerate(zip(paths, old_paths)):
+            if path != old_path:
+                new_paths.append(path)
+                old_videos.append(all_videos[video_idx])
 
-        return True
+        params["import_list"] = zip(ImportVideos().ask(filenames=new_paths), old_videos)
+
+        # This is the beginning of the "do" part of the funtion, but has a GUI pop-up.
+
+        def _trim_labeled_frames(lfs_to_remove):
+            """Trim labels past new video length."""
+            for lf in lfs_to_remove:
+                context.labels.remove_frame(lf)
+
+        def _reset_video_backend(video, filename, grayscale):
+            """Reset video back to original if operation aborted."""
+            video.backend.reset(
+                filename=filename,
+                grayscale=grayscale,
+            )
+
+        import_list = params["import_list"]
+
+        for import_item, video in import_list:
+            import_params = import_item["params"]
+            old_path = video.backend.filename
+            old_grayscale = video.backend.grayscale
+            new_path = import_params["filename"]
+
+            # TODO: Will need to create a new backend if import has different extension.
+            # Currently reset is only functional for MediaVideo backend.
+            # FIXME: Ideally we would only mess with the backend after getting approval
+            # ...but, use this to read in the number of frames. Could use cv2 here....
+            if new_path != video.backend.filename:
+                _reset_video_backend(video, new_path, import_params)
```

Comment:
We really shouldn't be changing the backend before getting approval from the user. To separate into `ask` and `do_action`, we could read the number of frames using cv2:
```py
reader = cv2.VideoCapture(new_path)
last_vid_frame = int(reader.get(cv2.CAP_PROP_FRAME_COUNT))
```

Reply:
I chose to go with the above recommendation and create a new reader for each video to (possibly) import.